### PR TITLE
handle C++ unordered_map

### PR DESCRIPTION
**Pull request by tgymnich**

### CODE REVIEW

File: enzyme/Enzyme/ActivityAnalysis.cpp [hunk 1]
```diff
@@ -84,65 +84,41 @@ cl::opt<bool>
...
+    "f90io", "$ss5print",
     "_ZTv0_n24_NSoD", //"1Ev, 0Ev
-    "_ZNSt16allocator_traitsISaIdEE10deallocate",
-    "_ZNSaIcED1Ev",
+    "_ZNSt16allocator_traitsISaIdEE10deallocate", "_ZNSaIcED1Ev",
     "_ZNSaIcEC1Ev",
 #if LLVM_VERSION_MAJOR <= 8
     // TODO this returns allocated memory and thus can be an active value
     // "_ZNSt16allocator_traits",
-    "_ZN4core3fmt",
```
Comment (wsmoses) [inline enzyme/Enzyme/ActivityAnalysis.cpp:96]: Fix newlines, then LGTM.

**Review by wsmoses — APPROVED** [review]

File: enzyme/Enzyme/ActivityAnalysis.cpp (hunk 1 above)
Comment (tgymnich) [reply enzyme/Enzyme/ActivityAnalysis.cpp]: fixed with `// clang-format on / off`

File: enzyme/Enzyme/TypeAnalysis/TypeAnalysis.cpp [hunk 2]
```diff
@@ -3761,6 +3762,17 @@ void TypeAnalyzer::visitCallInst(CallInst &call) {
...
       return;
     }
+
+    if (funcName.startswith("_ZNKSt3__14hash")) {
+      updateAnalysis(&call, TypeTree(BaseType::Integer).Only(-1, &call), &call);
+      return;
+    }
+
+    if (funcName.startswith("_ZNSt3__19addressof")) {
+      updateAnalysis(&call, TypeTree(BaseType::Pointer).Only(-1, &call), &call);
```
Comment (wsmoses) [inline enzyme/Enzyme/TypeAnalysis/TypeAnalysis.cpp:3772]: Why is this necessary?

Comment (tgymnich) [reply enzyme/Enzyme/TypeAnalysis/TypeAnalysis.cpp:3772]: crashes on mac os otherwise

Comment (wsmoses) [reply enzyme/Enzyme/TypeAnalysis/TypeAnalysis.cpp:3772]: Can you get an LL case of that?

Comment (tgymnich) [reply enzyme/Enzyme/TypeAnalysis/TypeAnalysis.cpp:3772]: https://fwd.gymni.ch/A5K94e

Comment (wsmoses) [reply enzyme/Enzyme/TypeAnalysis/TypeAnalysis.cpp:3772]: The problem actually appears to be a union in here:
```llvm
; Function Attrs: nounwind ssp willreturn mustprogress
define linkonce_odr hidden i8* @_ZNKSt3__112basic_stringIcNS_11char_traitsIcEENS_9allocatorIcEEE13__get_pointerEv(%"class.std::__1::basic_string"* nonnull dereferenceable(24) %this) local_unnamed_addr #9 align 2 {
entry:
  %call = call zeroext i1 @_ZNKSt3__112basic_stringIcNS_11char_traitsIcEENS_9allocatorIcEEE9__is_longEv(%"class.std::__1::basic_string"* nonnull dereferenceable(24) %this) #20
  br i1 %call, label %cond.true, label %cond.false

cond.true:                                        ; preds = %entry
  %call2 = call i8* @_ZNKSt3__112basic_stringIcNS_11char_traitsIcEENS_9allocatorIcEEE18__get_long_pointerEv(%"class.std::__1::basic_string"* nonnull dereferenceable(24) %this) #20
  br label %cond.end

cond.false:                                       ; preds = %entry
  %call3 = call i8* @_ZNKSt3__112basic_stringIcNS_11char_traitsIcEENS_9allocatorIcEEE19__get_short_pointerEv(%"class.std::__1::basic_string"* nonnull dereferenceable(24) %this) #20
  br label %cond.end

cond.end:                                         ; preds = %cond.false, %cond.true
  %cond = phi i8* [ %call2, %cond.true ], [ %call3, %cond.false ]
  ret i8* %cond
}
```

Comment (wsmoses) [reply enzyme/Enzyme/TypeAnalysis/TypeAnalysis.cpp:3772]: Can we just disable type analysis for this function?

Comment (tgymnich) [reply enzyme/Enzyme/TypeAnalysis/TypeAnalysis.cpp:3772]: I'll try

Comment (tgymnich) [reply enzyme/Enzyme/TypeAnalysis/TypeAnalysis.cpp:3774]: it works!
